src/auth/src/token_cache.rs

```diff
@@ -0,0 +1,594 @@
+        assert_eq!(actual, initial);
+
+        // wait long enough for the token to be expired
+        let sleep = TOKEN_VALID_DURATION.add(Duration::from_secs(100));
+        tokio::time::advance(sleep).await;
+
+        // make sure we return the error, not the expired token
+        assert!(cache.get_token().await.is_err());
+    }
+
+    #[tokio::test(flavor = "multi_thread", worker_threads = 8)]
+    async fn token_cache_multiple_requests_existing_valid_token() {
+        let now = Instant::now();
+
+        let token = Token {
+            token: "initial-token".to_string(),
+            token_type: "Bearer".to_string(),
+            expires_at: Some((now + TOKEN_VALID_DURATION).into_std()),
+            metadata: None,
+        };
+        let token_clone = token.clone();
+
+        let mut mock = MockTokenProvider::new();
+        mock.expect_get_token()
+            .times(1)
+            .return_once(|| Ok(token_clone));
+
+        // fetch an initial token
+        let cache = TokenCache::new(mock);
+        let actual = cache.get_token().await.unwrap();
+        assert_eq!(actual, token);
+
+        // Spawn N tasks, all asking for a token at once.
+        let tasks = (0..1000)
+            .map(|_| {
+                let cache_clone = cache.clone();
+                tokio::spawn(async move { cache_clone.get_token().await })
+            })
+            .collect::<Vec<_>>();
+
+        // Wait for the N token requests to complete, verifying the returned token.
+        for task in tasks {
+            let actual = task.await.unwrap();
+            assert!(actual.is_ok(), "{}", actual.err().unwrap());
+            assert_eq!(actual.unwrap(), token);
+        }
+    }
+
+    #[tokio::test]
+    async fn refresh_task_expired_token_loop() {
+        let now = Instant::now();
+
+        let token1 = Token {
+            token: "token1".to_string(),
+            token_type: "Bearer".to_string(),
+            expires_at: Some(now.into_std()),
+            metadata: None,
+        };
+        let token1_clone = token1.clone();
+
+        let token2 = Token {
+            token: "token2".to_string(),
+            token_type: "Bearer".to_string(),
+            expires_at: Some((now + TOKEN_VALID_DURATION).into_std()),
+            metadata: None,
+        };
+        let token2_clone = token2.clone();
+
+        let mut mock = MockTokenProvider::new();
+        mock.expect_get_token()
+            .times(1)
+            .return_once(|| Ok(token1_clone));
+
+        mock.expect_get_token()
+            .times(1)
+            .return_once(|| Ok(token2_clone));
+
+        let (tx, mut rx) = watch::channel::<Option<Result<Token>>>(None);
+
+        // Wait to make sure the first token expires.
+        sleep(Duration::from_millis(10)).await;
+
+        tokio::spawn(async move {
+            refresh_task(mock, tx).await;
+        });
+
+        // Give the refresh task a chance to run
+        sleep(Duration::from_millis(100)).await;
```

Comment:
These are race conditions and flakes waiting to happen. Can we write this test using `start_paused = true`?

Reply:
Removed one of the sleeps. Added it while debugging the issue (turned out to be somehow windows cannot give an instant in the past but windows and mac can). 

I do not think start-pasued=true is of use here because we do not want forward time. We want to make sure not much time passes between refreshes i.e. we are running the loop in refresh task without any sleep.

Reply:
Regardless, any test with a time in it is risky.  Maybe you will get lucky, but flaky tests are put out of their misery fast in this project.  I will not hesitate to remove or disable the test.